sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -212,6 +212,8 @@ class MicroBatchExecution(
     logInfo(s"Query $prettyIdString was stopped")
   }
 
+  private val watermarkPropagator = WatermarkPropagator(sparkSession.sessionState.conf)
```

Comment:
How do we recover `WatermarkPropagator` for a query restart?

Reply:
Whenever we can provide the origin watermark, it can deterministically simulate the watermark propagation. We can do this deterministically from the global watermark info in the offset metadata log/commit log.